src/main/java/com/iexec/core/task/TaskService.java

```diff
@@ -632,46 +635,41 @@ private void resultUploaded2Finalizing(Task task) {
             return;
         }
 
-        updateTaskStatusAndSave(task, FINALIZING);
-
-        Optional<ChainReceipt> optionalChainReceipt = iexecHubService.finalizeTask(task.getChainTaskId(), task.getResultLink(), task.getChainCallbackData());
-        if (!optionalChainReceipt.isPresent()) {
-            return;
-        }
-        ChainReceipt chainReceipt = optionalChainReceipt.get();
-
-        Optional<ChainTask> oChainTask = iexecHubService.getChainTask(task.getChainTaskId());
-        if (!oChainTask.isPresent()) {
-            return;
-        }
-        chainTask = oChainTask.get();
-
-        if (chainTask.getStatus().equals(ChainTaskStatus.FAILLED)) {
-            log.error("Finalize failed [chainTaskId:{} canFinalize:{}, isAfterRevealDeadline:{}, hasAtLeastOneReveal:{}]",
-                    task.getChainTaskId(), isTaskInResultUploaded, canFinalize, offChainRevealEqualsOnChainReveal);
-            updateTaskStatusAndSave(task, FINALIZE_FAILED);
-            updateTaskStatusAndSave(task, FAILED);
-            return;
-        }
-
-        finalizing2Finalized2Completed(task, chainReceipt);
+        blockchainAdapterService
+                .requestFinalize(task.getChainTaskId(), task.getResultLink(),
+                        task.getChainCallbackData())
+                .ifPresentOrElse(chainTaskId -> {
+                    log.info("Requested finalize on blockchain " +
+                            "[chainTaskId:{}]", task.getChainTaskId());
+                    updateTaskStatusAndSave(task, FINALIZING);
+                    //Watch finalizing to finalized
+                    updateTaskRunnable(task.getChainTaskId());
+                }, () -> {
+                    log.error("Failed to request finalize on blockchain " +
+                            "[chainTaskId:{}]", task.getChainTaskId());
+                    updateTaskStatusAndSave(task, FINALIZE_FAILED);
+                    updateTaskStatusAndSave(task, FAILED);
+                });
     }
 
     private void finalizing2Finalized2Completed(Task task) {
-        finalizing2Finalized2Completed(task, null);
-    }
-
-    private void finalizing2Finalized2Completed(Task task, ChainReceipt chainReceipt) {
-        Optional<ChainTask> oChainTask = iexecHubService.getChainTask(task.getChainTaskId());
-        if (!oChainTask.isPresent()) {
-            return;
-        }
-        ChainTask chainTask = oChainTask.get();
-
-        if (chainTask.getStatus().equals(ChainTaskStatus.COMPLETED)) {
-            updateTaskStatusAndSave(task, FINALIZED, chainReceipt);
-            finalizedToCompleted(task);
-        }
+        blockchainAdapterService
+                .isFinalized(task.getChainTaskId())
+                .ifPresentOrElse(isSuccess -> {
+                    if (isSuccess.equals(Boolean.TRUE)) {
```

Comment:
```suggestion
                    if (isSuccess) {
```
?

Reply:
Updated.
FYI:
I had a sonar warning just having  if (isSuccess)

Reply:
Ok, good to know then :+1: